File: internal/api/models/details.go
package models

import "time"

type EventDetails struct {
	ID                      string        `json:"id"`
	ChannelName             string        `json:"channel_name"`
	CreatedAt               time.Time     `json:"created_at"`
	PredictionWindowSeconds int           `json:"prediction_window_seconds"`
	Title                   string        `json:"title"`
	EventSeries             []EventSeries `json:"event_series"`
}

type EventSeries struct {
	ID        string           `json:"id"`
	Timestamp time.Time        `json:"timestamp"`
	Status    string           `json:"status"`
	Outcomes  []OutcomeDetails `json:"outcomes"`
}

type OutcomeDetails struct {
	ID            string      `json:"id"`
	Color         string      `json:"color"`
	Title         string      `json:"title"`
	BadgeVersion  string      `json:"badge_version"`
	BadgeSetID    string      `json:"badge_set_id"`
	Timestamp     time.Time   `json:"timestamp"`
	TotalPoints   int         `json:"total_points"`
	TotalUsers    int         `json:"total_users"`
	ResultType    string      `json:"result_type"`
	TopPredictors []Predictor `json:"top_predictors"`
}

type Predictor struct {
	User   User `json:"user"`
	Points int  `json:"points"`
}

File: ui/src/models/Details.ts
import Model from "./Model";
import Outcome from "./Outcome";
import { sort } from "@/utils/Timestamp";

interface Details {
  id: string;
  channel_name: string;
  created_at: string;
  prediction_window_seconds: number;
  title: string;
  event_series: EventSeries[];
}

interface EventSeries {
  id: string;
  timestamp: string;
  status: string;
  outcomes: Outcome[];
}

export interface DetailsSeries {
  details: Outcome;
  values: Array<{
    timestamp: Date;
    status: string;
    points: number;
    users: number;
  }>;
}

class Details extends Model {
  resource() {
    return "api/v1/details";
  }

  sort() {
    this.event_series.sort(sort);
  }

  getTimeSeries(): DetailsSeries[] {
    const outcomes: {
      [Key: string]: DetailsSeries;
    } = {};

    this.event_series.forEach((e) => {
      e.outcomes.forEach((o) => {
        if (!(o.title in outcomes)) {
          outcomes[o.title] = {
            details: o,
            values: [],
          };
        }

        const ts = new Date(e.timestamp);
        outcomes[o.title].values.push({
          timestamp: ts,
          status: e.status,
          points: o.total_points,
          users: o.total_users,
        });
      });
    });

    return Object.values(outcomes);
  }
}

export default Details;
